jvm-packages/xgboost4j-gpu/src/main/java/ml/dmlc/xgboost4j/gpu/java/CudfColumnBatch.java

```diff
@@ -0,0 +1,175 @@
+/*
+ Copyright (c) 2021 by Contributors
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+ http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+ */
+
+package ml.dmlc.xgboost4j.gpu.java;
+
+import java.util.Arrays;
+
+import ai.rapids.cudf.ColumnVector;
+import ai.rapids.cudf.Table;
+
+import ml.dmlc.xgboost4j.java.ColumnBatch;
+
+/**
+ * Class to wrap CUDF Table to generate the cuda array interface.
+ */
+public class CudfColumnBatch extends ColumnBatch {
+  private final Table table;          // The CUDF Table
+  private final int[] featureIndices; // The feature columns
+  private final int[] labelIndices;   // The label columns
+  private final int[] weightIndices;  // The weight columns
+  private final int[] baseMarginIndices; // The base margin columns
+
+  /**
+   * CudfColumnBatch constructor
+   * @param table             the CUDF table
+   * @param featureIndices    must-have, specify the feature's indices in the table
+   */
+  public CudfColumnBatch(Table table, int[] featureIndices) {
+    this(table, featureIndices, null, null, null);
+  }
+
+  /**
+   * CudfColumnBatch constructor
+   * @param table             the CUDF table
+   * @param featureIndices    must-have, specify the feature's indices in the table
+   * @param labelIndices      optional, specify the label's indices in the table
+   */
+  public CudfColumnBatch(Table table, int[] featureIndices, int[] labelIndices) {
+    this(table, featureIndices, labelIndices, null, null);
+  }
+
+  /**
+   * CudfColumnBatch constructor
+   * @param table             the CUDF table
+   * @param featureIndices    must-have, specify the feature's indices in the table
+   * @param labelIndices      must-have, specify the label's indices in the table
+   * @param weightIndices     optional, specify the weight's indices in the table
+   * @param baseMarginIndices optional, specify the base marge's indices in the table
+   */
+  public CudfColumnBatch(Table table, int[] featureIndices, int[] labelIndices, int[] weightIndices,
```

Comment:
I think this class should only contain feature columns, and the additional columns should be added like

> CudfColumn labelColumn = CudfColumn.from(tmpTable.getColumn(12));

Or is there performance reason why it was done like this?

Reply:
This class will be used to create DeviceDMatrix which requires feature/label (must-have) and weight/base_margin (optional)

Reply:
CudfColumnBatch will be used to construct Normal Dmatrix and DeviceDMatrix. For that purpose, I define 3 different constructors for it. the first one is only accepting features which will be used to create normal dmatrix. and the others are for creating DeviceDMatrix. So it doesn't matter here.

Reply:
Can you have a different constructor for DeviceDMatrix requiring these inputs upfront?

Reply:
Hi @RAMitchell, what does that mean?

Reply:
Here is what I imagine it could look like:
```
CudfColumnBatch batch = new CudfColumnBatch(tmpTable, new int[]{0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11});
CudfColumn labelColumn = CudfColumn.from(tmpTable.getColumn(12));

DMatrix dMatrix1 = new DMatrix(batch, Float.NaN, 1);
dMatrix1.setLabel(labelColumn)

// Require this class to include meta info vectors in constructor
DeviceDMatrix dMatrix2 = new DeviceDMatrix(batch, Float.NaN, 1, labelColumn, null, null);
```
Also, if you can do the slicing of tmpTable using native cudf functionality it's better.

e.g.
```
// CudfColumnBatch just accepts a dataframe and we use cudf to separate the parts we want
CudfColumnBatch batch = new CudfColumnBatch(tmpTable.getColumns(new int[]{0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11}));
```
I'm assuming here we can do this without actually copying any data, let me know if that assumption is incorrect.

Reply:
Is there any necessary to separate DeviceDmatrix from DMatrix? 

And BTW, there is not involving any copying data from GPU to CPU ...

Reply:
Remind me why does the DeviceDmatrix need the label column in the constructor and DMatrix doesn't?

Reply:
@RAMitchell I think `DeviceQuantileDMatrix` can only be used for training while `DMatrix` can be used for other puroses.

Reply:
Just to be clear, is `DeviceDmatrix` the same thing as `DeviceQuantileDMatrix` in this context?

> Is there any necessary to separate DeviceDmatrix from DMatrix?

If this achieves the goal of an interface that makes sense, then yes.